libyul/optimiser/StackCompressor.cpp

```diff
@@ -105,12 +114,40 @@ class RematCandidateSelector: public DataFlowAnalyzer
 		m_expressionCodeCost.erase(_variable);
 	}
 
+	/// All candidate variables in order of occurrence.
+	vector<YulString> m_candidates;
 	/// Candidate variables and the code cost of their value.
 	map<YulString, size_t> m_expressionCodeCost;
 	/// Number of references to each candidate variable.
 	map<YulString, size_t> m_numReferences;
 };
 
+/// Selects at most @a _numVariables among @a _candidates.
+set<YulString> chooseVarsToEliminate(
+	map<size_t, vector<tuple<YulString, set<YulString>>>> const& _candidates,
+	size_t _numVariables
+)
+{
+	set<YulString> varsToEliminate;
+	for (auto&& [cost, candidates]: _candidates)
+		for (auto&& [candidate, references]: candidates)
+		{
+			if (varsToEliminate.size() >= _numVariables)
+				return varsToEliminate;
+			// If a variable we would like to eliminate references another one
+			// we already selected for elimination, then stop selecting
+			// candidates. If we would add that variable, then the cost calculation
+			// for the previous variable would be off. Furthermore, we
+			// do not skip the variable because it would be better to properly re-compute
+			// the costs of all other variables instead.
+			for (YulString const& referencedVar: references)
+				if (varsToEliminate.count(referencedVar))
+					return varsToEliminate;
```

Comment:
Why are we returning early instead of continuing in the main loop like in the old implementation?

Reply:
The old implementation didn't continue for this case either, did it?
It had
```
		if (referencesVarToEliminate)
			break;
```
in the outer loop.

Reply:
Ah, right.

Reply:
The old implementation just didn't ``return``, because it wants to hit the part after the loop - which is why I extracted this to a function, s.t. we don't need these weird flags anymore and *can* just return from nested loops.